maintainers_annotation/maintainers_annotation.go

```diff
@@ -0,0 +1,190 @@
+// Copyright 2024 OpenSSF Scorecard Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package maintainers_annotation
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/ossf/scorecard/v4/checker"
+	"github.com/ossf/scorecard/v4/checks/fileparser"
+	"github.com/ossf/scorecard/v4/clients"
+	sce "github.com/ossf/scorecard/v4/errors"
+	"gopkg.in/yaml.v3"
+)
+
+var errNoScorecardYmlFile = errors.New("scorecard.yml doesn't exist or file doesn't contain exemptions")
+
+// AnnotationReason is the reason behind an annotation.
+type AnnotationReason string
+
+const (
+	// AnnotationTestData is to annotate when a check or probe is targeting a danger
+	// in files or code snippets only used for test or example purposes.
+	AnnotationTestData AnnotationReason = "test-data"
+	// AnnotationRemediated is to annotate when a check or probe correctly identified a
+	// danger and, even though the danger is necessary, a remediation was already applied.
+	// E.g. a workflow is dangerous but only run under maintainers verification and approval,
+	// or a binary is needed but it is signed or has provenance.
+	AnnotationRemediated AnnotationReason = "remediated"
+	// AnnotationNotApplicable is to annotate when a check or probe is not applicable for the case.
+	// E.g. the dependencies should not be pinned because the project is a library.
+	AnnotationNotApplicable AnnotationReason = "not-applicable"
+	// AnnotationNotSupported is to annotate when the maintainer fulfills a check or probe in a way
+	// that is not supported by Scorecard. E.g. Clang-Tidy is used as SAST tool but not identified
+	// because its not supported.
+	AnnotationNotSupported AnnotationReason = "not-supported"
+	// AnnotationNotDetected is to annotate when the maintainer fulfills a check or probe in a way
+	// that is supported by Scorecard but not identified. E.g. Dependabot is configured in the
+	// repository settings and not in a file.
+	AnnotationNotDetected AnnotationReason = "not-detected"
+)
+
+// AnnotationReasonDoc maps the reasons to their human-readable explanations.
+var AnnotationReasonDoc = map[AnnotationReason]string{
```

Comment:
public, maps are generally discouraged, as they're world-writable.

Reply:
Another way to do this is:
```golang
type AnnotationReason struct {
  Label string
  Doc string
}
```

Reply:
Changed to a switch approach instead of putting it in a struct. In the struct approach I understand I would still need to instantiate a global obj and switches seemed more performatic. Let me know if you agree with the new approach.